File: src/syntax/token.ts
export interface Token {
  type: TType
  value?: string
  line: number
  span: {
    start: number
    offset: number
  }
}

export function defaultToken(): Token {
  return {
    type: tokens.eof, line: 0, span: { start: 0, offset: 0 }
  }
}

export interface TType {
  label: string
  operator?: boolean
  prefix?: boolean
  precedence?: number
}

function make(label: string, operator?: boolean, precedence?: number, prefix?: boolean) : TType {
  return {
    label, operator: !!operator, prefix: !!prefix, precedence
  }
}

export const tokens = {
  lparen: make('('),
  rparen: make(')'),
  lbrace: make('{'),
  rbrace: make('}'),
  lbracket: make('['),
  rbracket: make(']'),
  dot: make('.'),
  comma: make(','),
  colon: make(':'),
  hash: make('#'),
  semicolon: make(';'),
  underscore: make('_'),
  arrow: make('->'),
  equals: make('='),
  walrus: make(':='),
  
  mut: make('mut'),
  if:  make('if'),
  then: make('then'),
  else: make('else'),

  mul: make('*', true, 7),
  div: make('/', true, 7),
  mod: make('%', true, 7),
  add: make('+', true, 6),
  sub: make('-', true, 6),
  lt: make('<', true, 5),
  le: make('<=', true, 5),
  gt: make('>', true, 5),
  ge: make('>=', true, 5),
  ne: make('/=', true, 5),
  eq: make('==', true, 5),
  and: make('and', true, 4),
  or: make('or', true, 3),
  is: make('is', true, 2),
  pipe: make('|>', true, 1),

  not: make('not', true, 0, true),
  
  lower: make('lowercase name'),
  upper: make('uppercase name'),
  number: make('number'),
  string: make('string'),
  
  stringStart: make('start of a template'),
  stringPart: make('part of a template'),
  stringFinish: make('end of a template'),

  eof: make('eof'),
}

export const keywords: { [key: string]: TType } = {
  'and': tokens.and,
  'or': tokens.or,
  'is': tokens.is,
  'not': tokens.not,
  'if': tokens.if,
  'then': tokens.then,
  'else': tokens.else,
  'mut': tokens.mut,
}

export const symbols: { [key: string]: TType } = {
  ':': tokens.colon,
  '.': tokens.dot,
  '->': tokens.arrow,
  '=': tokens.equals,
  ':=': tokens.walrus,
  '*': tokens.mul,
  '/': tokens.div,
  '%': tokens.mod,
  '+': tokens.add,
  '-': tokens.sub,
  '<': tokens.lt,
  '<=': tokens.le,
  '>': tokens.gt,
  '>=': tokens.ge,
  '/=': tokens.ne,
  '==': tokens.eq,
  '|>': tokens.pipe,
}

File: src/syntax/token.rs
use std::fmt;

use crate::span::Span;

#[derive(Debug, Clone, Copy, Eq, PartialEq, Default)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

#[derive(Debug, Clone, Copy, Eq, PartialEq, Default)]
pub enum TokenKind {
    #[default]
    EOF,
    NL,

    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,

    Dot,    // .
    Comma,  // ,
    Semi,   // ;
    Colon,  // :
    Equals, // =
    Arrow,  // =>

    And,
    Or,
    Not,
    Is,
    Mut,
    If,
    Else,
    For,
    In,
    Case,
    Where,
    Import,
    Yield,
    Break,
    Return,

    Plus,    // +
    Minus,   // -
    Star,    // *
    Slash,   // /
    Mod,     // %
    Caret,   // ^
    EqEq,    // ==
    NoEq,    // !=
    Lt,      // <
    LtEq,    // <=
    Gt,      // >
    GtEq,    // >=
    Dots,    // ..

    Ident,
    Number,
    Integer,
    String,

    UnexpectedCharacter,
    UnterminatedString,
    Noop,
}

impl fmt::Display for TokenKind {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let description = match self {
            TokenKind::EOF => "end of input",
            TokenKind::NL => "new line",
            TokenKind::LParen => "(",
            TokenKind::RParen => ")",
            TokenKind::LBrace => "{",
            TokenKind::RBrace => "}",
            TokenKind::LBracket => "[",
            TokenKind::RBracket => "]",
            TokenKind::Dot => ".",
            TokenKind::Comma => ",",
            TokenKind::Colon => ":",
            TokenKind::Semi => ";",
            TokenKind::Equals => "=",
            TokenKind::Arrow => "=>",
            TokenKind::And => "and",
            TokenKind::Or => "or",
            TokenKind::Not => "not",
            TokenKind::Is => "is",
            TokenKind::Mut => "mut",
            TokenKind::If => "if",
            TokenKind::Else => "else",
            TokenKind::For => "for",
            TokenKind::In => "in",
            TokenKind::Case => "case",
            TokenKind::Where => "where",
            TokenKind::Import => "import",
            TokenKind::Yield => "yield",
            TokenKind::Break => "break",
            TokenKind::Return => "return",
            TokenKind::Plus => "+",
            TokenKind::Minus => "-",
            TokenKind::Star => "*",
            TokenKind::Slash => "/",
            TokenKind::Caret => "^",
            TokenKind::Mod => "%",
            TokenKind::EqEq => "==",
            TokenKind::NoEq => "!=",
            TokenKind::Lt => "<",
            TokenKind::LtEq => "<=",
            TokenKind::Gt => ">",
            TokenKind::GtEq => ">=",
            TokenKind::Dots => "..",
            TokenKind::Ident => "identifier",
            TokenKind::Number => "number",
            TokenKind::Integer => "integer",
            TokenKind::String => "string",
            TokenKind::UnexpectedCharacter => "unexpected character",
            TokenKind::UnterminatedString => "unterminated string",
            TokenKind::Noop => "placeholder",
        };
        write!(f, "{description}")
    }
}

impl Token {
    pub fn is(self, kind: TokenKind) -> bool {
        self.kind == kind
    }

    pub fn non_terminal(self) -> bool {
        matches!(
            self.kind,
            TokenKind::Dot
          | TokenKind::Plus
          | TokenKind::Minus
          | TokenKind::Star
          | TokenKind::Slash
          | TokenKind::Mod
          | TokenKind::Caret
          | TokenKind::EqEq
          | TokenKind::NoEq
          | TokenKind::Lt
          | TokenKind::LtEq
          | TokenKind::Gt
          | TokenKind::GtEq
          | TokenKind::Dots
        )
    }
}
